<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.27">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<title>Exercise 2 â€“ Biology-informed Integration and Visualization of Multiomics Data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../assets/SIB_logo.svg" rel="icon" type="image/svg+xml">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-ed96de9b727972fe78a7b5d16c58bf87.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-416d0afab5eaf77b6e3abee7d4113ad9.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script><link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Source+Sans+Pro">
<!-- Cookie Consent & Matomo Tracking --><script src="assets/cookie-consent.js"></script><!-- End Matomo Code -->
</head>
<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="navbar navbar-expand-lg " data-bs-theme="dark"><div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Biology-informed Integration and Visualization of Multiomics Data</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
<li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../precourse_preparations.html"> 
<span class="menu-text">Precourse preparations</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../course_schedule.html"> 
<span class="menu-text">Course schedule</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-1" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">day 1</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-1">
<li>
    <a class="dropdown-item" href="../day1/day1-1_dataset.html">
 <span class="dropdown-text">Exercise 1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day1/day1-2_atac.html">
 <span class="dropdown-text">Exercise 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day1/day1-3_chip_rna.html">
 <span class="dropdown-text">Exercise 3</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-2" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">day 2</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-2">
<li>
    <a class="dropdown-item" href="../day2/day2-1_overlap_matrix.html">
 <span class="dropdown-text">Exercise 4</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day2/day2-2_atac_rna.html">
 <span class="dropdown-text">Exercise 5</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day2/day2-3_multiomics.html">
 <span class="dropdown-text">Exercise 6</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-3" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">day 3</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-3">
<li>
    <a class="dropdown-item" href="../day3/day3-1_enrichment_analysis.html">
 <span class="dropdown-text">Exercise 7</span></a>
  </li>  
    </ul>
</li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-projects" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">projects</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-projects">
<li>
    <a class="dropdown-item" href="../projects/introduction.html">
 <span class="dropdown-text">Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../projects/project1.html">
 <span class="dropdown-text">Project 1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../projects/project2.html">
 <span class="dropdown-text">Project 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../projects/project3.html">
 <span class="dropdown-text">Project 3</span></a>
  </li>  
    </ul>
</li>
</ul>
<ul class="navbar-nav navbar-nav-scroll ms-auto">
<li class="nav-item compact">
    <a class="nav-link" href="https://github.com/sib-swiss/multiomics-biological-integration-training"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
</div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Exercise 2</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../index.html" class="sidebar-logo-link">
      <img src="../assets/SIB_LogoQ_GBv.svg" alt="" class="sidebar-logo light-content py-0 d-lg-inline d-none"><img src="../assets/SIB_LogoQ_GBv.svg" alt="" class="sidebar-logo dark-content py-0 d-lg-inline d-none"></a>
      </div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li>
<a href="#enriched-heatmaps-and-identification-of-regions-with-different-activities" id="toc-enriched-heatmaps-and-identification-of-regions-with-different-activities" class="nav-link active" data-scroll-target="#enriched-heatmaps-and-identification-of-regions-with-different-activities">Enriched Heatmaps and identification of regions with different activities</a>
  <ul class="collapse">
<li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#load-libraries" id="toc-load-libraries" class="nav-link" data-scroll-target="#load-libraries">Load Libraries</a></li>
  <li><a href="#atac-se-object" id="toc-atac-se-object" class="nav-link" data-scroll-target="#atac-se-object">ATAC SE object</a></li>
  <li><a href="#atac-peaks-info" id="toc-atac-peaks-info" class="nav-link" data-scroll-target="#atac-peaks-info">ATAC peaks info</a></li>
  <li><a href="#volcanoplot-to-observe-differentially-accessible-peaks" id="toc-volcanoplot-to-observe-differentially-accessible-peaks" class="nav-link" data-scroll-target="#volcanoplot-to-observe-differentially-accessible-peaks">Volcanoplot to observe differentially accessible peaks</a></li>
  <li><a href="#taking-1000-bp-around-the-mid-of-atac-peaks" id="toc-taking-1000-bp-around-the-mid-of-atac-peaks" class="nav-link" data-scroll-target="#taking-1000-bp-around-the-mid-of-atac-peaks">Taking 1000 bp around the mid of ATAC-peaks</a></li>
  <li><a href="#normalizing-data-for-plotting" id="toc-normalizing-data-for-plotting" class="nav-link" data-scroll-target="#normalizing-data-for-plotting">Normalizing data for plotting</a></li>
  <li><a href="#enriched-heatmap" id="toc-enriched-heatmap" class="nav-link" data-scroll-target="#enriched-heatmap">Enriched heatmap</a></li>
  <li>
<a href="#changing-aesthectics-of-enriched-heatmap" id="toc-changing-aesthectics-of-enriched-heatmap" class="nav-link" data-scroll-target="#changing-aesthectics-of-enriched-heatmap">Changing aesthectics of Enriched heatmap</a>
  <ul class="collapse">
<li><a href="#changind-color-and-size" id="toc-changind-color-and-size" class="nav-link" data-scroll-target="#changind-color-and-size">Changind color and size</a></li>
  <li><a href="#color-based-on-quantile" id="toc-color-based-on-quantile" class="nav-link" data-scroll-target="#color-based-on-quantile">Color based on quantile</a></li>
  <li><a href="#changing-some-other-aesthetics" id="toc-changing-some-other-aesthetics" class="nav-link" data-scroll-target="#changing-some-other-aesthetics">Changing some other aesthetics</a></li>
  <li><a href="#split-the-enriched-heatmap-based-on-logfc-values" id="toc-split-the-enriched-heatmap-based-on-logfc-values" class="nav-link" data-scroll-target="#split-the-enriched-heatmap-based-on-logfc-values">Split the Enriched heatmap based on <code>logFC</code> values</a></li>
  <li><a href="#make-a-function-to-make-this-heatmap" id="toc-make-a-function-to-make-this-heatmap" class="nav-link" data-scroll-target="#make-a-function-to-make-this-heatmap">Make a function to make this heatmap</a></li>
  <li><a href="#make-enriched-heatmaps-for-both-atac-samples" id="toc-make-enriched-heatmaps-for-both-atac-samples" class="nav-link" data-scroll-target="#make-enriched-heatmaps-for-both-atac-samples">Make Enriched Heatmaps for both ATAC samples</a></li>
  </ul>
</li>
  <li><a href="#another-way-to-make-annotations-for-split" id="toc-another-way-to-make-annotations-for-split" class="nav-link" data-scroll-target="#another-way-to-make-annotations-for-split">Another way to make annotations for split</a></li>
  <li><a href="#question-1" id="toc-question-1" class="nav-link" data-scroll-target="#question-1">Question 1</a></li>
  <li><a href="#question-2" id="toc-question-2" class="nav-link" data-scroll-target="#question-2">Question 2</a></li>
  <li><a href="#question-3" id="toc-question-3" class="nav-link" data-scroll-target="#question-3">Question 3</a></li>
  </ul>
</li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">Exercise 2</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header><section id="enriched-heatmaps-and-identification-of-regions-with-different-activities" class="level1"><h1>Enriched Heatmaps and identification of regions with different activities</h1>
<section id="learning-objectives" class="level2"><h2 class="anchored" data-anchor-id="learning-objectives">Learning Objectives</h2>
<ul>
<li>By the end of this section, you will be able to:</li>
<li>Distinguish between when to use <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap()</a></code> and <code><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation()</a></code>
</li>
<li>Generate row split annotations for enriched heatmaps</li>
<li>Add multiple categorical annotations with separate color legends</li>
<li>Avoid common data handling errors in <code>heatmap</code> plotting</li>
</ul></section><section id="load-libraries" class="level2"><h2 class="anchored" data-anchor-id="load-libraries">Load Libraries</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/EnrichedHeatmap">EnrichedHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">gUtils</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">rtracklayer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/circlize">circlize</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="atac-se-object" class="level2"><h2 class="anchored" data-anchor-id="atac-se-object">ATAC SE object</h2>
<p>Here we load the <code>SummarizedExperiment</code> for the ATAC-seq data</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">atac</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/atac_se.rds"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="atac-peaks-info" class="level2"><h2 class="anchored" data-anchor-id="atac-peaks-info">ATAC peaks info</h2>
<p>Each peak region represent the activity in a genomic region. These peaks are annotated for various features, including their genomic annotations and the results for differential accessibility analysis is also added there.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb3"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">atac</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3656 ranges and 18 metadata columns:
         seqnames              ranges strand |             annotation   geneChr
            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |            &lt;character&gt; &lt;integer&gt;
     [1]     chr1     3670547-3672665      * |       Promoter (&lt;=1kb)         1
     [2]     chr1     4332510-4332710      * | Intron (ENSMUST00000..         1
     [3]     chr1     4491755-4492573      * |       Promoter (1-2kb)         1
     [4]     chr1     4571186-4572423      * |      Distal Intergenic         1
     [5]     chr1     4785062-4786325      * |       Promoter (&lt;=1kb)         1
     ...      ...                 ...    ... .                    ...       ...
  [3652]     chr2 181764664-181764891      * |       Promoter (1-2kb)         2
  [3653]     chr2 181766881-181767875      * |       Promoter (&lt;=1kb)         2
  [3654]     chr2 181837244-181838655      * |       Promoter (&lt;=1kb)         2
  [3655]     chr2 181863853-181865066      * |       Promoter (&lt;=1kb)         2
  [3656]     chr2 181918525-181918853      * |      Distal Intergenic         2
         geneStart   geneEnd geneLength geneStrand      geneId
         &lt;integer&gt; &lt;integer&gt;  &lt;integer&gt;  &lt;integer&gt; &lt;character&gt;
     [1]   3214482   3671498     457017          2      497097
     [2]   4344146   4360314      16169          2       19888
     [3]   4492465   4493735       1271          2       20671
     [4]   4491390   4497354       5965          2       20671
     [5]   4773206   4785710      12505          2       27395
     ...       ...       ...        ...        ...         ...
  [3652] 181763332 181827797      64466          1       17932
  [3653] 181767029 181795892      28864          1       17932
  [3654] 181837854 181857461      19608          1      245867
  [3655] 181864337 181870830       6494          1       67005
  [3656] 181864360 181866209       1850          1       67005
                  transcriptId distanceToTSS            ENSEMBL      SYMBOL
                   &lt;character&gt;     &lt;numeric&gt;        &lt;character&gt; &lt;character&gt;
     [1]  ENSMUST00000070533.4             0 ENSMUSG00000051951        Xkr4
     [2]  ENSMUST00000027032.5         27604 ENSMUSG00000025900         Rp1
     [3]  ENSMUST00000191939.1          1162 ENSMUSG00000025902       Sox17
     [4]  ENSMUST00000192650.5        -73832 ENSMUSG00000025902       Sox17
     [5]  ENSMUST00000130201.7             0 ENSMUSG00000033845      Mrpl15
     ...                   ...           ...                ...         ...
  [3652] ENSMUST00000081125.10          1332 ENSMUSG00000010505        Myt1
  [3653]  ENSMUST00000156190.7             0 ENSMUSG00000010505        Myt1
  [3654] ENSMUST00000029116.13             0 ENSMUSG00000027589      Pcmtd2
  [3655]  ENSMUST00000039551.8             0 ENSMUSG00000038628      Polr3k
  [3656]  ENSMUST00000153214.1         54165 ENSMUSG00000038628      Polr3k
                       GENENAME        anno     logFC    logCPM         F
                    &lt;character&gt; &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
     [1] X-linked Kx blood gr..    Promoter  0.602984  10.45378  41.78446
     [2] retinitis pigmentosa..      Intron  0.855748   5.18424   3.68371
     [3] SRY (sex determining..    Promoter -0.620926   7.13469   9.14522
     [4] SRY (sex determining..  Intergenic -0.511046   8.90885  15.15688
     [5] mitochondrial riboso..    Promoter  0.540908   8.63469  15.08892
     ...                    ...         ...       ...       ...       ...
  [3652] myelin transcription..    Promoter -1.275275   5.54135  10.44866
  [3653] myelin transcription..    Promoter  0.477658   7.56969   6.03877
  [3654] protein-L-isoasparta..    Promoter  0.689247   7.92175  16.67881
  [3655] polymerase (RNA) III..    Promoter  0.241693   9.05746   4.06676
  [3656] polymerase (RNA) III..  Intergenic  0.394577   6.84745   2.73331
              PValue      qvalue
           &lt;numeric&gt;   &lt;numeric&gt;
     [1] 6.58453e-08 4.81983e-07
     [2] 6.13517e-02 9.28785e-02
     [3] 4.12368e-03 8.38540e-03
     [4] 3.27158e-04 8.67115e-04
     [5] 3.36058e-04 8.82634e-04
     ...         ...         ...
  [3652] 0.002309221  0.00497558
  [3653] 0.017945118  0.03116850
  [3654] 0.000181039  0.00051468
  [3655] 0.049773014  0.07677364
  [3656] 0.105289776  0.15077925
  -------
  seqinfo: 21 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">rd_atac</span> <span class="op">&lt;-</span> <span class="fu">rowRanges</span><span class="op">(</span><span class="va">atac</span><span class="op">)</span></span>
<span><span class="va">rd_atac</span> <span class="op">&lt;-</span> <span class="va">rd_atac</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">rd_atac</span><span class="op">$</span><span class="va">logFC</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">0.5</span> <span class="op">&amp;</span> <span class="va">rd_atac</span><span class="op">$</span><span class="va">qvalue</span> <span class="op">&lt;=</span> <span class="fl">0.1</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">elementMetadata</a></span><span class="op">(</span><span class="va">rd_atac</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="st">"ATAC"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html">elementMetadata</a></span><span class="op">(</span><span class="va">rd_atac</span><span class="op">)</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="volcanoplot-to-observe-differentially-accessible-peaks" class="level2"><h2 class="anchored" data-anchor-id="volcanoplot-to-observe-differentially-accessible-peaks">Volcanoplot to observe differentially accessible peaks</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span><span class="va">rd_atac</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  seqnames   start     end width strand
1     chr1 3670547 3672665  2119      *
2     chr1 4332510 4332710   201      *
3     chr1 4491755 4492573   819      *
4     chr1 4571186 4572423  1238      *
5     chr1 4785062 4786325  1264      *
6     chr1 5232000 5233386  1387      *
                                      ATAC_annotation ATAC_geneChr
1                                    Promoter (&lt;=1kb)            1
2 Intron (ENSMUST00000208660.1/19888, intron 3 of 29)            1
3                                    Promoter (1-2kb)            1
4                                   Distal Intergenic            1
5                                    Promoter (&lt;=1kb)            1
6                                   Distal Intergenic            1
  ATAC_geneStart ATAC_geneEnd ATAC_geneLength ATAC_geneStrand ATAC_geneId
1        3214482      3671498          457017               2      497097
2        4344146      4360314           16169               2       19888
3        4492465      4493735            1271               2       20671
4        4491390      4497354            5965               2       20671
5        4773206      4785710           12505               2       27395
6        5083164      5099777           16614               1      108664
     ATAC_transcriptId ATAC_distanceToTSS       ATAC_ENSEMBL ATAC_SYMBOL
1 ENSMUST00000070533.4                  0 ENSMUSG00000051951        Xkr4
2 ENSMUST00000027032.5              27604 ENSMUSG00000025900         Rp1
3 ENSMUST00000191939.1               1162 ENSMUSG00000025902       Sox17
4 ENSMUST00000192650.5             -73832 ENSMUSG00000025902       Sox17
5 ENSMUST00000130201.7                  0 ENSMUSG00000033845      Mrpl15
6 ENSMUST00000192142.1             148836 ENSMUSG00000033793     Atp6v1h
                                    ATAC_GENENAME  ATAC_anno ATAC_logFC
1               X-linked Kx blood group related 4   Promoter  0.6029841
2                  retinitis pigmentosa 1 (human)     Intron  0.8557477
3           SRY (sex determining region Y)-box 17   Promoter -0.6209259
4           SRY (sex determining region Y)-box 17 Intergenic -0.5110459
5             mitochondrial ribosomal protein L15   Promoter  0.5409079
6 ATPase, H+ transporting, lysosomal V1 subunit H Intergenic -2.5017799
  ATAC_logCPM     ATAC_F  ATAC_PValue  ATAC_qvalue
1   10.453783  41.784460 6.584528e-08 4.819830e-07
2    5.184242   3.683714 6.135165e-02 9.287853e-02
3    7.134690   9.145224 4.123677e-03 8.385399e-03
4    8.908847  15.156883 3.271580e-04 8.671152e-04
5    8.634686  15.088922 3.360576e-04 8.826341e-04
6    8.555241 257.544409 3.686136e-20 3.369128e-17</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html">rownames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="st">"region"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html">nrow</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>         seqnames   start     end width strand
region_1     chr1 3670547 3672665  2119      *
region_2     chr1 4332510 4332710   201      *
region_3     chr1 4491755 4492573   819      *
region_4     chr1 4571186 4572423  1238      *
region_5     chr1 4785062 4786325  1264      *
region_6     chr1 5232000 5233386  1387      *
                                             ATAC_annotation ATAC_geneChr
region_1                                    Promoter (&lt;=1kb)            1
region_2 Intron (ENSMUST00000208660.1/19888, intron 3 of 29)            1
region_3                                    Promoter (1-2kb)            1
region_4                                   Distal Intergenic            1
region_5                                    Promoter (&lt;=1kb)            1
region_6                                   Distal Intergenic            1
         ATAC_geneStart ATAC_geneEnd ATAC_geneLength ATAC_geneStrand
region_1        3214482      3671498          457017               2
region_2        4344146      4360314           16169               2
region_3        4492465      4493735            1271               2
region_4        4491390      4497354            5965               2
region_5        4773206      4785710           12505               2
region_6        5083164      5099777           16614               1
         ATAC_geneId    ATAC_transcriptId ATAC_distanceToTSS       ATAC_ENSEMBL
region_1      497097 ENSMUST00000070533.4                  0 ENSMUSG00000051951
region_2       19888 ENSMUST00000027032.5              27604 ENSMUSG00000025900
region_3       20671 ENSMUST00000191939.1               1162 ENSMUSG00000025902
region_4       20671 ENSMUST00000192650.5             -73832 ENSMUSG00000025902
region_5       27395 ENSMUST00000130201.7                  0 ENSMUSG00000033845
region_6      108664 ENSMUST00000192142.1             148836 ENSMUSG00000033793
         ATAC_SYMBOL                                   ATAC_GENENAME  ATAC_anno
region_1        Xkr4               X-linked Kx blood group related 4   Promoter
region_2         Rp1                  retinitis pigmentosa 1 (human)     Intron
region_3       Sox17           SRY (sex determining region Y)-box 17   Promoter
region_4       Sox17           SRY (sex determining region Y)-box 17 Intergenic
region_5      Mrpl15             mitochondrial ribosomal protein L15   Promoter
region_6     Atp6v1h ATPase, H+ transporting, lysosomal V1 subunit H Intergenic
         ATAC_logFC ATAC_logCPM     ATAC_F  ATAC_PValue  ATAC_qvalue
region_1  0.6029841   10.453783  41.784460 6.584528e-08 4.819830e-07
region_2  0.8557477    5.184242   3.683714 6.135165e-02 9.287853e-02
region_3 -0.6209259    7.134690   9.145224 4.123677e-03 8.385399e-03
region_4 -0.5110459    8.908847  15.156883 3.271580e-04 8.671152e-04
region_5  0.5409079    8.634686  15.088922 3.360576e-04 8.826341e-04
region_6 -2.5017799    8.555241 257.544409 3.686136e-20 3.369128e-17</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Define thresholds</span></span>
<span><span class="va">fc_cutoff</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span><span class="va">q_cutoff</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span></span>
<span><span class="co"># Count significant peaks</span></span>
<span><span class="va">up_sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">ATAC_logFC</span> <span class="op">&gt;</span> <span class="va">fc_cutoff</span> <span class="op">&amp;</span> <span class="va">df</span><span class="op">$</span><span class="va">ATAC_qvalue</span> <span class="op">&lt;</span> <span class="va">q_cutoff</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">down_sig</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">ATAC_logFC</span> <span class="op">&lt;</span> <span class="op">-</span><span class="va">fc_cutoff</span> <span class="op">&amp;</span> <span class="va">df</span><span class="op">$</span><span class="va">ATAC_qvalue</span> <span class="op">&lt;</span> <span class="va">q_cutoff</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ATAC_logFC</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">ATAC_qvalue</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">ATAC_logFC</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"TRUE"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"FALSE"</span> <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">xlab</a></span><span class="op">(</span><span class="st">"Log2 Fold Change"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ylab</a></span><span class="op">(</span><span class="st">"-Log10 Q-value"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Volcano Plot of Differentially Accessible Peaks"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="va">fc_cutoff</span>, <span class="va">fc_cutoff</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">q_cutoff</span><span class="op">)</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span>, color <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotate.html">annotate</a></span><span class="op">(</span><span class="st">"text"</span>, x <span class="op">=</span> <span class="fl">2</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">ATAC_qvalue</span><span class="op">)</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>           label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Up: "</span>, <span class="va">up_sig</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"red"</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/annotate.html">annotate</a></span><span class="op">(</span><span class="st">"text"</span>, x <span class="op">=</span> <span class="op">-</span><span class="fl">2</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">ATAC_qvalue</span><span class="op">)</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>           label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"Down: "</span>, <span class="va">down_sig</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"blue"</span>, hjust <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="day1-2_atac_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="taking-1000-bp-around-the-mid-of-atac-peaks" class="level2"><h2 class="anchored" data-anchor-id="taking-1000-bp-around-the-mid-of-atac-peaks">Taking 1000 bp around the mid of ATAC-peaks</h2>
<p>Most of the peaks are around 1000 bp wide. We can check that</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html">lengths</a></span><span class="op">(</span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">atac</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1031</code></pre>
</div>
</div>
<p>For plotting the data, we can hence consider 1000 bp around the <strong>peak-mid</strong></p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Finding mid of all peaks</span></span>
<span><span class="va">mid_peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/gUtils/man/gr.mid.html">gr.mid</a></span><span class="op">(</span><span class="va">rd_atac</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># adding names to peaks to give them identity</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mid_peaks</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="st">"peak"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">mid_peaks</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="normalizing-data-for-plotting" class="level2"><h2 class="anchored" data-anchor-id="normalizing-data-for-plotting">Normalizing data for plotting</h2>
<p><code>bigWig</code> files are compressed, indexed, binary format used for efficiently displaying continuous data, like genomic signal data, in genome browsers. Here, we read in ATAC-seq <code>bigWig</code> files, filters the data to specific chromosomes, normalizes signal intensity around genomic regions of interest (peak centers), and saves the resulting matrices for downstream visualization.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb14"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">atac_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html">list.files</a></span><span class="op">(</span><span class="st">"data"</span>, pattern <span class="op">=</span> <span class="st">"ATAC"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">atac_files</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"\\.bw"</span>, replacement <span class="op">=</span> <span class="st">""</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html">basename</a></span><span class="op">(</span><span class="va">atac_files</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">atac_bw</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html">lapply</a></span><span class="op">(</span><span class="va">atac_files</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">a</span> <span class="op">&lt;-</span> <span class="fu">rtracklayer</span><span class="fu">::</span><span class="fu">import</span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>  <span class="va">a</span> <span class="op">&lt;-</span> <span class="va">a</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/Seqinfo/man/seqinfo.html">seqnames</a></span><span class="op">(</span><span class="va">a</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">a</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<p><code>bigWig</code> files are represented as <code>GRanges</code>.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">atac_bw</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>$ATAC_11half
GRanges object with 6354084 ranges and 1 metadata column:
            seqnames              ranges strand |     score
               &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
        [1]     chr1           1-3000000      * | 0.0000000
        [2]     chr1     3000001-3000050      * | 0.0225005
        [3]     chr1     3000051-3000100      * | 0.0600013
        [4]     chr1     3000101-3000150      * | 0.0337507
        [5]     chr1     3000151-3000200      * | 0.0375008
        ...      ...                 ...    ... .       ...
  [6354080]     chr2 182013051-182013100      * |   45.3122
  [6354081]     chr2 182013101-182013150      * |   49.6173
  [6354082]     chr2 182013151-182013200      * |   79.4379
  [6354083]     chr2 182013201-182013250      * |   50.2623
  [6354084]     chr2 182013251-182113224      * |    0.0000
  -------
  seqinfo: 66 sequences from an unspecified genome

$ATAC_15half
GRanges object with 5356375 ranges and 1 metadata column:
            seqnames              ranges strand |      score
               &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;numeric&gt;
        [1]     chr1           1-3000400      * | 0.00000000
        [2]     chr1     3000401-3000500      * | 0.00844927
        [3]     chr1     3000501-3000550      * | 0.00000000
        [4]     chr1     3000551-3000650      * | 0.00844927
        [5]     chr1     3000651-3003300      * | 0.00000000
        ...      ...                 ...    ... .        ...
  [5356371]     chr2 182012501-182012600      * | 0.01689850
  [5356372]     chr2 182012601-182012650      * | 0.00844927
  [5356373]     chr2 182012651-182012900      * | 0.00000000
  [5356374]     chr2 182012901-182013000      * | 0.01689850
  [5356375]     chr2 182013001-182113224      * | 0.00000000
  -------
  seqinfo: 66 sequences from an unspecified genome</code></pre>
</div>
</div>
<p>Next, we calculate the normalized signals into the area of our interest. Please check <code><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">?normalizeToMatrix</a></code> for details of this function</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mat_AS</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html">lapply</a></span><span class="op">(</span><span class="va">atac_bw</span>, FUN <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/normalizeToMatrix.html">normalizeToMatrix</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">mid_peaks</span>,</span>
<span>    extend <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    value_column <span class="op">=</span> <span class="st">"score"</span>,</span>
<span>    include_target <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    mean_mode <span class="op">=</span> <span class="st">"w0"</span>,</span>
<span>    w <span class="op">=</span> <span class="fl">20</span>,</span>
<span>    smooth <span class="op">=</span> <span class="cn">T</span>,</span>
<span>    background <span class="op">=</span> <span class="fl">0</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mat_AS</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>$ATAC_11half
Normalize x to mid_peaks:
  Upstream 1000 bp (50 windows)
  Downstream 1000 bp (50 windows)
  Include target regions (width = 1)
  1933 target regions

$ATAC_15half
Normalize x to mid_peaks:
  Upstream 1000 bp (50 windows)
  Downstream 1000 bp (50 windows)
  Include target regions (width = 1)
  1933 target regions</code></pre>
</div>
</div>
</section><section id="enriched-heatmap" class="level2"><h2 class="anchored" data-anchor-id="enriched-heatmap">Enriched heatmap</h2>
<p><strong>Enriched heatmap</strong> is a special type of heatmap which visualizes the enrichment of genomic signals over specific target regions.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="va">mat_AS</span><span class="op">$</span><span class="va">ATAC_11half</span>, name <span class="op">=</span> <span class="st">"E11.5"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span><span class="op">(</span>mat <span class="op">=</span> <span class="va">mat_AS</span><span class="op">$</span><span class="va">ATAC_15half</span>, name <span class="op">=</span> <span class="st">"E15.5"</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>The automatically generated colors map from the 1^st and 99^th of the
values in the matrix. There are outliers in the matrix whose patterns
might be hidden by this color mapping. You can manually set the color
to `col` argument.

Use `suppressMessages()` to turn off this message.
The automatically generated colors map from the 1^st and 99^th of the
values in the matrix. There are outliers in the matrix whose patterns
might be hidden by this color mapping. You can manually set the color
to `col` argument.

Use `suppressMessages()` to turn off this message.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="day1-2_atac_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Joining 2 <code>EnrichedHeatmaps</code> is very easy with a <code>+</code> sign.</p>
</section><section id="changing-aesthectics-of-enriched-heatmap" class="level2"><h2 class="anchored" data-anchor-id="changing-aesthectics-of-enriched-heatmap">Changing aesthectics of Enriched heatmap</h2>
<p>Letâ€™s work on one data for now.</p>
<section id="changind-color-and-size" class="level3"><h3 class="anchored" data-anchor-id="changind-color-and-size">Changind color and size</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span><span class="op">(</span></span>
<span>  mat <span class="op">=</span> <span class="va">mat_AS</span><span class="op">$</span><span class="va">ATAC_11half</span>,    <span class="co"># normalized matrix</span></span>
<span>  name <span class="op">=</span> <span class="st">"E11.5"</span>,              <span class="co"># Name for the plot</span></span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span>,     <span class="co"># We change the colors for low to high values</span></span>
<span>  width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"cm"</span><span class="op">)</span>,       <span class="co"># Width of the heatmap</span></span>
<span>  height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">8</span>, <span class="st">"cm"</span><span class="op">)</span>      <span class="co"># Height of the heatmap</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="day1-2_atac_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You may wonder why the color looks so light. The reason is in coverage values in ATAC, there exist some extreme values, which results in extreme value in <code>normalizedMatrix</code>.</p>
</section><section id="color-based-on-quantile" class="level3"><h3 class="anchored" data-anchor-id="color-based-on-quantile">Color based on quantile</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Taking data between 1 and 99 percentile</span></span>
<span><span class="va">col_fun</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">mat_AS</span><span class="op">$</span><span class="va">ATAC_11half</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.99</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span><span class="op">(</span></span>
<span>  mat <span class="op">=</span> <span class="va">mat_AS</span><span class="op">$</span><span class="va">ATAC_11half</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"E11.5"</span>,</span>
<span>  col <span class="op">=</span> <span class="va">col_fun</span>,</span>
<span>  width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">8</span>, <span class="st">"cm"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="day1-2_atac_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="changing-some-other-aesthetics" class="level3"><h3 class="anchored" data-anchor-id="changing-some-other-aesthetics">Changing some other aesthetics</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb23"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># We first change the color legent of the plot to show only 3 values</span></span>
<span><span class="va">vmin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">mat_AS</span><span class="op">$</span><span class="va">ATAC_11half</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.01</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vmax</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">mat_AS</span><span class="op">$</span><span class="va">ATAC_11half</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.99</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vmid</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">vmin</span> <span class="op">+</span> <span class="va">vmax</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span></span>
<span><span class="va">legend_ticks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">vmin</span>, <span class="va">vmid</span>, <span class="va">vmax</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span><span class="op">(</span></span>
<span>  mat <span class="op">=</span> <span class="va">mat_AS</span><span class="op">$</span><span class="va">ATAC_11half</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"E11.5"</span>,</span>
<span>  col <span class="op">=</span> <span class="va">col_fun</span>,</span>
<span>  width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">8</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  column_title <span class="op">=</span> <span class="st">"E11.5"</span>,</span>
<span>  column_title_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">10</span>, fill <span class="op">=</span> <span class="st">"#ffcccc"</span><span class="op">)</span>,</span>
<span>  axis_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"-1kb"</span>, <span class="st">"mid"</span>, <span class="st">"1kb"</span><span class="op">)</span>,  <span class="co"># We changed the axis names here</span></span>
<span> heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    at <span class="op">=</span> <span class="va">legend_ticks</span>,</span>
<span>    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">legend_ticks</span>, digits <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>    title_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>    labels_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  top_annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span></span>
<span>    lines <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span><span class="op">(</span></span>
<span>      height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>      gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span></span>
<span>        lwd <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>        fontsize <span class="op">=</span> <span class="fl">5</span></span>
<span>      <span class="op">)</span>,</span>
<span>      axis_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        side <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>        facing <span class="op">=</span> <span class="st">"inside"</span>,</span>
<span>        gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span></span>
<span>          fontsize <span class="op">=</span> <span class="fl">7</span>,</span>
<span>          col <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>          lwd <span class="op">=</span> <span class="fl">0.4</span></span>
<span>        <span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="day1-2_atac_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="split-the-enriched-heatmap-based-on-logfc-values" class="level3"><h3 class="anchored" data-anchor-id="split-the-enriched-heatmap-based-on-logfc-values">Split the Enriched heatmap based on <code>logFC</code> values</h3>
<p>Although we see some signal here, but it might be a good idea to split the heatmap into the regions which gained and lost accessibility.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">split_change</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">mid_peaks</span><span class="op">$</span><span class="va">ATAC_logFC</span> <span class="op">&gt;</span> <span class="fl">0</span>, yes <span class="op">=</span> <span class="st">"Increased accessibility"</span>, no <span class="op">=</span> <span class="st">"Decreased accessibility"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">split_change</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mid_peaks</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">split_change</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   peak_1                    peak_2                    peak_3 
"Increased accessibility" "Increased accessibility" "Decreased accessibility" 
                   peak_4                    peak_5                    peak_6 
"Decreased accessibility" "Increased accessibility" "Decreased accessibility" </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb26"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Define cluster colors</span></span>
<span><span class="va">cluster_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Increased accessibility"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"Decreased accessibility"</span> <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make sure split_change has levels matching the color names</span></span>
<span><span class="va">split_change</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="va">split_change</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cluster_colors</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span><span class="op">(</span></span>
<span>  mat <span class="op">=</span> <span class="va">mat_AS</span><span class="op">$</span><span class="va">ATAC_11half</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"E11.5"</span>,</span>
<span>  row_split <span class="op">=</span> <span class="va">split_change</span>,</span>
<span>  col <span class="op">=</span> <span class="va">col_fun</span>,</span>
<span>  width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">8</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  column_title <span class="op">=</span> <span class="st">"E11.5"</span>,</span>
<span>  column_title_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">10</span>, fill <span class="op">=</span> <span class="st">"#ffcccc"</span><span class="op">)</span>,</span>
<span>  axis_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"-1kb"</span>, <span class="st">"mid"</span>, <span class="st">"1kb"</span><span class="op">)</span>,</span>
<span>  heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    at <span class="op">=</span> <span class="va">legend_ticks</span>,</span>
<span>    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">legend_ticks</span>, digits <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>    title_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>    labels_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  top_annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span></span>
<span>    lines <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span><span class="op">(</span></span>
<span>      gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">cluster_colors</span><span class="op">)</span>,</span>
<span>      height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>      axis_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        side <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>        facing <span class="op">=</span> <span class="st">"inside"</span>,</span>
<span>        gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span></span>
<span>          fontsize <span class="op">=</span> <span class="fl">7</span>,</span>
<span>          lwd <span class="op">=</span> <span class="fl">0.4</span></span>
<span>        <span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="day1-2_atac_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="make-a-function-to-make-this-heatmap" class="level3"><h3 class="anchored" data-anchor-id="make-a-function-to-make-this-heatmap">Make a function to make this heatmap</h3>
<p>As you know we have at least 2 samples as of now. It will be a good idea to create a <code>function</code> to make this heatmap.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">make_EH</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">norm_mat</span>, <span class="va">heatmap_cols</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"white"</span>, <span class="st">"red"</span><span class="op">)</span>, <span class="va">split_rows</span> <span class="op">=</span> <span class="cn">NULL</span>, <span class="va">hm_name</span>, <span class="va">col_fill</span> <span class="op">=</span> <span class="st">"#ffcccc"</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">col_fun</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/circlize/man/colorRamp2.html">colorRamp2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">norm_mat</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.01</span>, <span class="fl">0.99</span><span class="op">)</span><span class="op">)</span>, <span class="va">heatmap_cols</span><span class="op">)</span>  </span>
<span>  </span>
<span>  <span class="va">vmin</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">norm_mat</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.01</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">vmax</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/quantile.html">quantile</a></span><span class="op">(</span><span class="va">norm_mat</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0.99</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">vmid</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">vmin</span> <span class="op">+</span> <span class="va">vmax</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span></span>
<span>  <span class="va">legend_ticks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">vmin</span>, <span class="va">vmid</span>, <span class="va">vmax</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/EnrichedHeatmap.html">EnrichedHeatmap</a></span><span class="op">(</span></span>
<span>  mat <span class="op">=</span> <span class="va">norm_mat</span>,</span>
<span>  name <span class="op">=</span> <span class="va">hm_name</span>,</span>
<span>  row_split <span class="op">=</span> <span class="va">split_rows</span>,</span>
<span>  col <span class="op">=</span> <span class="va">col_fun</span>,</span>
<span>  width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">4</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">8</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  column_title <span class="op">=</span> <span class="va">hm_name</span>,</span>
<span>  column_title_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">10</span>, fill <span class="op">=</span> <span class="va">col_fill</span><span class="op">)</span>,</span>
<span>  axis_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"-1kb"</span>, <span class="st">"mid"</span>, <span class="st">"1kb"</span><span class="op">)</span>,</span>
<span>  heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>    at <span class="op">=</span> <span class="va">legend_ticks</span>,</span>
<span>    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">legend_ticks</span>, digits <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>    title_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>    labels_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  top_annotation <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/HeatmapAnnotation.html">HeatmapAnnotation</a></span><span class="op">(</span></span>
<span>    lines <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/EnrichedHeatmap/man/anno_enriched.html">anno_enriched</a></span><span class="op">(</span></span>
<span>      gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>col <span class="op">=</span> <span class="va">cluster_colors</span><span class="op">)</span>,</span>
<span>      height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">2</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>      axis_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>        side <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>        facing <span class="op">=</span> <span class="st">"inside"</span>,</span>
<span>        gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span></span>
<span>          fontsize <span class="op">=</span> <span class="fl">7</span>,</span>
<span>          lwd <span class="op">=</span> <span class="fl">0.4</span></span>
<span>        <span class="op">)</span></span>
<span>      <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section><section id="make-enriched-heatmaps-for-both-atac-samples" class="level3"><h3 class="anchored" data-anchor-id="make-enriched-heatmaps-for-both-atac-samples">Make Enriched Heatmaps for both ATAC samples</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">eh_11h</span> <span class="op">&lt;-</span> <span class="fu">make_EH</span><span class="op">(</span>norm_mat <span class="op">=</span> <span class="va">mat_AS</span><span class="op">$</span><span class="va">ATAC_11half</span>, split_rows <span class="op">=</span> <span class="va">split_change</span>, hm_name <span class="op">=</span> <span class="st">"E11.5"</span><span class="op">)</span></span>
<span><span class="va">eh_15h</span> <span class="op">&lt;-</span> <span class="fu">make_EH</span><span class="op">(</span>norm_mat <span class="op">=</span> <span class="va">mat_AS</span><span class="op">$</span><span class="va">ATAC_15half</span>, split_rows <span class="op">=</span> <span class="va">split_change</span>, hm_name <span class="op">=</span> <span class="st">"E15.5"</span>, col_fill <span class="op">=</span> <span class="st">"#e6fff2"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="va">eh_11h</span> <span class="op">+</span> <span class="va">eh_15h</span>, merge_legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="day1-2_atac_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section></section><section id="another-way-to-make-annotations-for-split" class="level2"><h2 class="anchored" data-anchor-id="another-way-to-make-annotations-for-split">Another way to make annotations for split</h2>
<p>It is probably a good idea to represent clusters with colors, instead of text</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb29"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">eh_11h</span> <span class="op">&lt;-</span> <span class="fu">make_EH</span><span class="op">(</span>norm_mat <span class="op">=</span> <span class="va">mat_AS</span><span class="op">$</span><span class="va">ATAC_11half</span>, hm_name <span class="op">=</span> <span class="st">"E11.5"</span><span class="op">)</span></span>
<span><span class="va">eh_15h</span> <span class="op">&lt;-</span> <span class="fu">make_EH</span><span class="op">(</span>norm_mat <span class="op">=</span> <span class="va">mat_AS</span><span class="op">$</span><span class="va">ATAC_15half</span>, hm_name <span class="op">=</span> <span class="st">"E15.5"</span>, col_fill <span class="op">=</span> <span class="st">"#e6fff2"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">row_order_eh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/row_order-dispatch.html">row_order</a></span><span class="op">(</span><span class="va">eh_11h</span><span class="op">)</span></span>
<span></span>
<span><span class="va">anno_hm</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">split_change</span>,</span>
<span>  col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"blue"</span><span class="op">)</span>, </span>
<span>  name <span class="op">=</span> <span class="st">"Change"</span>,</span>
<span>  show_row_names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  show_column_names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">3</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>  height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">8</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  row_order <span class="op">=</span> <span class="va">row_order_eh</span>,</span>
<span>  row_title_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="va">anno_hm</span> <span class="op">+</span> <span class="va">eh_11h</span> <span class="op">+</span> <span class="va">eh_15h</span>, split <span class="op">=</span> <span class="va">split_change</span>, merge_legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="day1-2_atac_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section id="question-1" class="level2"><h2 class="anchored" data-anchor-id="question-1">Question 1</h2>
<p><strong>Can you make split the Enriched Heatmap based on the annotations?</strong></p>
<p><strong><em>Hint:</em></strong> <code>mid_peaks$ATAC_anno</code> contain annotations for the regions.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Tip</span>Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">split_anno</span> <span class="op">&lt;-</span> <span class="va">mid_peaks</span><span class="op">$</span><span class="va">ATAC_anno</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">split_anno</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mid_peaks</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">split_anno</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>      peak_1       peak_2       peak_3       peak_4       peak_5       peak_6 
  "Promoter"     "Intron"   "Promoter" "Intergenic"   "Promoter" "Intergenic" </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cols_an</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">split_anno</span><span class="op">)</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">anno_an</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">split_anno</span>,</span>
<span>  col <span class="op">=</span> <span class="va">cols_an</span>, </span>
<span>  name <span class="op">=</span> <span class="st">"anno"</span>,</span>
<span>  show_row_names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  show_column_names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">3</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>  height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">8</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  row_order <span class="op">=</span> <span class="va">row_order_eh</span>,</span>
<span>  row_title_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="va">anno_an</span> <span class="op">+</span> <span class="va">eh_11h</span> <span class="op">+</span> <span class="va">eh_15h</span>, split <span class="op">=</span> <span class="va">split_anno</span>, merge_legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="day1-2_atac_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="question-2" class="level2"><h2 class="anchored" data-anchor-id="question-2">Question 2</h2>
<p><strong>Can you make split the Enriched Heatmap based on the annotations and change in direction?</strong></p>
<p><strong><em>Hint:</em></strong> <code>mid_peaks$ATAC_anno</code> contain annotations for the regions. <code>mid_peaks$ATAC_logFC</code> contain sign of change.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-2-contents" aria-controls="callout-2" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Tip</span>Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-2" class="callout-2-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb33"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">split_anno_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html">paste</a></span><span class="op">(</span><span class="va">mid_peaks</span><span class="op">$</span><span class="va">ATAC_anno</span>, <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">mid_peaks</span><span class="op">$</span><span class="va">ATAC_logFC</span> <span class="op">&gt;</span> <span class="fl">0</span>, yes <span class="op">=</span> <span class="st">"Inc"</span>, no <span class="op">=</span> <span class="st">"Dec"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">split_anno_dir</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mid_peaks</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">split_anno_dir</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>          peak_1           peak_2           peak_3           peak_4 
  "Promoter Inc"     "Intron Inc"   "Promoter Dec" "Intergenic Dec" 
          peak_5           peak_6 
  "Promoter Inc" "Intergenic Dec" </code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cols_an</span> <span class="op">&lt;-</span> <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">split_anno_dir</span><span class="op">)</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"Paired"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">anno_an_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">split_anno_dir</span>,</span>
<span>  col <span class="op">=</span> <span class="va">cols_an</span>, </span>
<span>  name <span class="op">=</span> <span class="st">"anno"</span>,</span>
<span>  show_row_names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  show_column_names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">3</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>  height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">8</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  row_order <span class="op">=</span> <span class="va">row_order_eh</span>,</span>
<span>  row_title_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>There are 10 unique colors in the vector `col` and 10 unique values in
`matrix`. `Heatmap()` will treat it as an exact discrete one-to-one
mapping. If this is not what you want, slightly change the number of
colors, e.g. by adding one more color or removing a color.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb37"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="va">anno_an_dir</span> <span class="op">+</span> <span class="va">eh_11h</span> <span class="op">+</span> <span class="va">eh_15h</span>, split <span class="op">=</span> <span class="va">split_anno_dir</span>, merge_legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="day1-2_atac_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="question-3" class="level2"><h2 class="anchored" data-anchor-id="question-3">Question 3</h2>
<p><strong>Can you make split the Enriched Heatmap based on the annotations and change in direction with separate color bars for annotation and direction of change?</strong></p>
<p><strong><em>Hint:</em></strong> <code>mid_peaks$ATAC_anno</code> contain annotations for the regions. <code>mid_peaks$ATAC_logFC</code> contain sign of change.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-3-contents" aria-controls="callout-3" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
<span class="screen-reader-only">Tip</span>Answer
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-3" class="callout-3-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">split_anno_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span></span>
<span>  Annotation <span class="op">=</span> <span class="va">mid_peaks</span><span class="op">$</span><span class="va">ATAC_anno</span>,</span>
<span>  Direction <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">mid_peaks</span><span class="op">$</span><span class="va">ATAC_logFC</span> <span class="op">&gt;</span> <span class="fl">0</span>, yes <span class="op">=</span> <span class="st">"Inc"</span>, no <span class="op">=</span> <span class="st">"Dec"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">split_anno_df</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Annotation Direction
1   Promoter       Inc
2     Intron       Inc
3   Promoter       Dec
4 Intergenic       Dec
5   Promoter       Inc
6 Intergenic       Dec</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">cols_an</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"blue"</span>, </span>
<span>             <span class="fu">RColorBrewer</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">split_anno_df</span><span class="op">$</span><span class="va">Annotation</span><span class="op">)</span><span class="op">)</span>, name <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cols_an</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">split_anno_df</span><span class="op">$</span><span class="va">Direction</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unique.html">unique</a></span><span class="op">(</span><span class="va">split_anno_df</span><span class="op">$</span><span class="va">Annotation</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">anno_an_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html">Heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">split_anno_df</span>,</span>
<span>  name <span class="op">=</span> <span class="st">"anno"</span>,</span>
<span>  col <span class="op">=</span> <span class="va">cols_an</span>,</span>
<span>  show_row_names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  show_column_names <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>  width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">3</span>, <span class="st">"mm"</span><span class="op">)</span>,</span>
<span>  height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html">unit</a></span><span class="op">(</span><span class="fl">8</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>  row_order <span class="op">=</span> <span class="va">row_order_eh</span>,</span>
<span>  row_title_gp <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/gpar.html">gpar</a></span><span class="op">(</span>fontsize <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html">draw</a></span><span class="op">(</span><span class="va">anno_an_df</span> <span class="op">+</span> <span class="va">eh_11h</span> <span class="op">+</span> <span class="va">eh_15h</span>, split <span class="op">=</span> <span class="va">split_anno_dir</span>, merge_legend <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output-display">
<div>
<figure class="figure"><p><img src="day1-2_atac_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p><strong><code>Heatmap</code> and <code>EnrichedHeatmap</code> serve different functions, however the plots can be combined effortlessly. This makes the visualization of complex data easy.</strong></p>
</div>
</div>


</section></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "î§‹";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/sib-swiss\.github\.io\/multiomics-biological-integration-training\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->



</body></html>