<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.8.27">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Exercise 4 – Biology-informed Integration and Visualization of Multiomics Data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../assets/SIB_logo.svg" rel="icon" type="image/svg+xml">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/axe/axe-check.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-ed96de9b727972fe78a7b5d16c58bf87.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-416d0afab5eaf77b6e3abee7d4113ad9.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Source+Sans+Pro">
<!-- Cookie Consent & Matomo Tracking -->
<script src="assets/cookie-consent.js"></script>
<!-- End Matomo Code -->


</head>

<body class="nav-sidebar floating nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Biology-informed Integration and Visualization of Multiomics Data</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../precourse_preparations.html"> 
<span class="menu-text">Precourse preparations</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../course_schedule.html"> 
<span class="menu-text">Course schedule</span></a>
  </li>  
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-1" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">day 1</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-1">    
        <li>
    <a class="dropdown-item" href="../day1/day1-1_dataset.html">
 <span class="dropdown-text">Exercise 1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day1/day1-2_atac.html">
 <span class="dropdown-text">Exercise 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day1/day1-3_chip_rna.html">
 <span class="dropdown-text">Exercise 3</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-2" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">day 2</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-2">    
        <li>
    <a class="dropdown-item" href="../day2/day2-1_overlap_matrix.html">
 <span class="dropdown-text">Exercise 4</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day2/day2-2_atac_rna.html">
 <span class="dropdown-text">Exercise 5</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../day2/day2-3_multiomics.html">
 <span class="dropdown-text">Exercise 6</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-day-3" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">day 3</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-day-3">    
        <li>
    <a class="dropdown-item" href="../day3/day3-1_enrichment_analysis.html">
 <span class="dropdown-text">Exercise 7</span></a>
  </li>  
    </ul>
  </li>
  <li class="nav-item dropdown ">
    <a class="nav-link dropdown-toggle" href="#" id="nav-menu-projects" role="link" data-bs-toggle="dropdown" aria-expanded="false">
 <span class="menu-text">projects</span>
    </a>
    <ul class="dropdown-menu" aria-labelledby="nav-menu-projects">    
        <li>
    <a class="dropdown-item" href="../projects/introduction.html">
 <span class="dropdown-text">Introduction</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../projects/project1.html">
 <span class="dropdown-text">Project 1</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../projects/project2.html">
 <span class="dropdown-text">Project 2</span></a>
  </li>  
        <li>
    <a class="dropdown-item" href="../projects/project3.html">
 <span class="dropdown-text">Project 3</span></a>
  </li>  
    </ul>
  </li>
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/sib-swiss/multiomics-biological-integration-training"> <i class="bi bi-github" role="img" aria-label="GitHub">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">Exercise 4</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
      <a href="../index.html" class="sidebar-logo-link">
      <img src="../assets/SIB_LogoQ_GBv.svg" alt="" class="sidebar-logo light-content py-0 d-lg-inline d-none">
      <img src="../assets/SIB_LogoQ_GBv.svg" alt="" class="sidebar-logo dark-content py-0 d-lg-inline d-none">
      </a>
      </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#overlapping-all-the-datasets" id="toc-overlapping-all-the-datasets" class="nav-link active" data-scroll-target="#overlapping-all-the-datasets">Overlapping all the datasets</a>
  <ul class="collapse">
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#load-libraries" id="toc-load-libraries" class="nav-link" data-scroll-target="#load-libraries">Load Libraries</a></li>
  <li><a href="#se-objects" id="toc-se-objects" class="nav-link" data-scroll-target="#se-objects">SE objects</a></li>
  <li><a href="#overlap-matrix" id="toc-overlap-matrix" class="nav-link" data-scroll-target="#overlap-matrix">Overlap matrix</a></li>
  <li><a href="#function-to-merge-genomic-ranges" id="toc-function-to-merge-genomic-ranges" class="nav-link" data-scroll-target="#function-to-merge-genomic-ranges">Function to merge Genomic Ranges</a></li>
  <li><a href="#atac-seq" id="toc-atac-seq" class="nav-link" data-scroll-target="#atac-seq">ATAC-Seq</a></li>
  <li><a href="#rna" id="toc-rna" class="nav-link" data-scroll-target="#rna">RNA</a></li>
  <li><a href="#chip-seq" id="toc-chip-seq" class="nav-link" data-scroll-target="#chip-seq">ChIP-Seq</a>
  <ul class="collapse">
  <li><a href="#h3k4me3" id="toc-h3k4me3" class="nav-link" data-scroll-target="#h3k4me3">H3K4me3</a></li>
  <li><a href="#h3k4me1" id="toc-h3k4me1" class="nav-link" data-scroll-target="#h3k4me1">H3K4me1</a></li>
  <li><a href="#h3k27me3" id="toc-h3k27me3" class="nav-link" data-scroll-target="#h3k27me3">H3K27me3</a></li>
  <li><a href="#h3k27ac" id="toc-h3k27ac" class="nav-link" data-scroll-target="#h3k27ac">H3K27ac</a></li>
  </ul></li>
  <li><a href="#save-data" id="toc-save-data" class="nav-link" data-scroll-target="#save-data">Save data</a>
  <ul class="collapse">
  <li><a href="#snapshot-of-overlamatrix" id="toc-snapshot-of-overlamatrix" class="nav-link" data-scroll-target="#snapshot-of-overlamatrix">Snapshot of overlaMatrix</a></li>
  </ul></li>
  </ul></li>
  <li><a href="#normalized-metrics-for-plotting" id="toc-normalized-metrics-for-plotting" class="nav-link" data-scroll-target="#normalized-metrics-for-plotting">Normalized metrics for plotting</a>
  <ul class="collapse">
  <li><a href="#taking-1000-bp-around-the-mid-of-atac-peaks" id="toc-taking-1000-bp-around-the-mid-of-atac-peaks" class="nav-link" data-scroll-target="#taking-1000-bp-around-the-mid-of-atac-peaks">Taking 1000 bp around the mid of ATAC-peaks</a></li>
  <li><a href="#atac" id="toc-atac" class="nav-link" data-scroll-target="#atac">ATAC</a></li>
  <li><a href="#chip" id="toc-chip" class="nav-link" data-scroll-target="#chip">ChIP</a>
  <ul class="collapse">
  <li><a href="#h3k4me3-1" id="toc-h3k4me3-1" class="nav-link" data-scroll-target="#h3k4me3-1">H3K4me3</a></li>
  <li><a href="#h3k4me1-1" id="toc-h3k4me1-1" class="nav-link" data-scroll-target="#h3k4me1-1">H3K4me1</a></li>
  <li><a href="#h27k4me3" id="toc-h27k4me3" class="nav-link" data-scroll-target="#h27k4me3">H27K4me3</a></li>
  <li><a href="#h3k27ac-1" id="toc-h3k27ac-1" class="nav-link" data-scroll-target="#h3k27ac-1">H3K27ac</a></li>
  </ul></li>
  <li><a href="#rna-1" id="toc-rna-1" class="nav-link" data-scroll-target="#rna-1">RNA</a></li>
  <li><a href="#wgbs" id="toc-wgbs" class="nav-link" data-scroll-target="#wgbs">WGBS</a></li>
  <li><a href="#question-1" id="toc-question-1" class="nav-link" data-scroll-target="#question-1">Question 1</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Exercise 4</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="overlapping-all-the-datasets" class="level1">
<h1>Overlapping all the datasets</h1>
<section id="learning-objectives" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives">Learning Objectives</h2>
<p>By the end of this exercise, you will be able to: - Load and explore genomic data using Bioconductor classes such as <code>GRangesList</code> and <code>RangedSummarizedExperiment</code>. - Subset and filter genomic features based on annotation or genomic coordinates. - Perform basic overlap and distance-based queries between genomic intervals. - Integrate multi-assay data (e.g., chromatin marks and gene expression) using shared genomic coordinates.</p>
</section>
<section id="load-libraries" class="level2">
<h2 class="anchored" data-anchor-id="load-libraries">Load Libraries</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ComplexHeatmap)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnrichedHeatmap)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SummarizedExperiment)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gUtils)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rtracklayer)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(R.utils)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(circlize)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="se-objects" class="level2">
<h2 class="anchored" data-anchor-id="se-objects">SE objects</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>atac <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/atac_se.rds"</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>rna <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/rna_se.rds"</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>h3k4me3 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/h3k4me3_se.rds"</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>h3k4me1 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/h3k4me1_se.rds"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>h3k27me3 <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/h3k27me3_se.rds"</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>h3k27ac <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"data/h3k27ac_se.rds"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="overlap-matrix" class="level2">
<h2 class="anchored" data-anchor-id="overlap-matrix">Overlap matrix</h2>
<p>To make the <code>overlapMatrix</code> (you could also say <code>overlapRanges</code>), we need to identify a target.</p>
</section>
<section id="function-to-merge-genomic-ranges" class="level2">
<h2 class="anchored" data-anchor-id="function-to-merge-genomic-ranges">Function to merge Genomic Ranges</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge metadata from two GRanges objects based on overlaps</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>metaGR <span class="ot">&lt;-</span> <span class="cf">function</span>(gr1, gr2, <span class="at">minOverlap =</span> <span class="dv">1</span>) {</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Initialize metadata with gr1's metadata</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  mcols_out <span class="ot">&lt;-</span> <span class="fu">mcols</span>(gr1)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Find overlaps (keep all from gr1)</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>  hits <span class="ot">&lt;-</span> <span class="fu">findOverlaps</span>(gr1, gr2, <span class="at">minoverlap =</span> minOverlap)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  idx1 <span class="ot">&lt;-</span> <span class="fu">subjectHits</span>(hits)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>  idx2 <span class="ot">&lt;-</span> <span class="fu">queryHits</span>(hits)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Prepare new metadata to be added</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>  new_mcols <span class="ot">&lt;-</span> <span class="fu">DataFrame</span>(<span class="fu">matrix</span>(<span class="cn">NA</span>, <span class="at">nrow =</span> <span class="fu">length</span>(gr1), <span class="at">ncol =</span> <span class="fu">ncol</span>(<span class="fu">mcols</span>(gr2))))</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(new_mcols) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(<span class="fu">mcols</span>(gr2))</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Fill metadata only for overlaps</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>  new_mcols[idx2, ] <span class="ot">&lt;-</span> <span class="fu">mcols</span>(gr2)[idx1, , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Combine original and new metadata</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mcols</span>(gr1) <span class="ot">&lt;-</span> <span class="fu">cbind</span>(mcols_out, new_mcols)</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(gr1)</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="atac-seq" class="level2">
<h2 class="anchored" data-anchor-id="atac-seq">ATAC-Seq</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract rowRanges and filter peaks with strong signal</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>rd_atac <span class="ot">&lt;-</span> <span class="fu">rowRanges</span>(atac)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>rd_atac <span class="ot">&lt;-</span> rd_atac[<span class="fu">abs</span>(rd_atac<span class="sc">$</span>logFC) <span class="sc">&gt;=</span> <span class="fl">0.5</span> <span class="sc">&amp;</span> rd_atac<span class="sc">$</span>qvalue <span class="sc">&lt;=</span> <span class="fl">0.1</span>]</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(<span class="fu">elementMetadata</span>(rd_atac)) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"ATAC"</span>, <span class="fu">colnames</span>(<span class="fu">elementMetadata</span>(rd_atac)), <span class="at">sep =</span> <span class="st">"_"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="rna" class="level2">
<h2 class="anchored" data-anchor-id="rna">RNA</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter differentially expressed genes</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>rd_rna <span class="ot">&lt;-</span> <span class="fu">rowRanges</span>(rna)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>rd_rna <span class="ot">&lt;-</span> rd_rna[<span class="fu">abs</span>(rd_rna<span class="sc">$</span>logFC) <span class="sc">&gt;=</span> <span class="fl">0.5</span> <span class="sc">&amp;</span> rd_rna<span class="sc">$</span>qvalue <span class="sc">&lt;=</span> <span class="fl">0.1</span>]</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(<span class="fu">elementMetadata</span>(rd_rna)) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"RNA"</span>, <span class="fu">colnames</span>(<span class="fu">elementMetadata</span>(rd_rna)), <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>overlap <span class="ot">&lt;-</span> <span class="fu">metaGR</span>(<span class="at">gr1 =</span> rd_atac, <span class="at">gr2 =</span> rd_rna, <span class="at">minOverlap =</span> <span class="dv">10</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="chip-seq" class="level2">
<h2 class="anchored" data-anchor-id="chip-seq">ChIP-Seq</h2>
<section id="h3k4me3" class="level3">
<h3 class="anchored" data-anchor-id="h3k4me3">H3K4me3</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>rd_h3k4me3 <span class="ot">&lt;-</span> <span class="fu">rowRanges</span>(h3k4me3)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>rd_h3k4me3 <span class="ot">&lt;-</span> rd_h3k4me3[<span class="fu">abs</span>(rd_h3k4me3<span class="sc">$</span>logFC) <span class="sc">&gt;=</span> <span class="fl">0.5</span> <span class="sc">&amp;</span> rd_h3k4me3<span class="sc">$</span>qvalue <span class="sc">&lt;=</span> <span class="fl">0.1</span>]</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(<span class="fu">elementMetadata</span>(rd_h3k4me3)) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"H3K4me3"</span>, <span class="fu">colnames</span>(<span class="fu">elementMetadata</span>(rd_h3k4me3)), <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>overlap <span class="ot">&lt;-</span> <span class="fu">metaGR</span>(<span class="at">gr1 =</span> overlap, <span class="at">gr2 =</span> rd_h3k4me3, <span class="at">minOverlap =</span> <span class="dv">10</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="h3k4me1" class="level3">
<h3 class="anchored" data-anchor-id="h3k4me1">H3K4me1</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>rd_h3k4me1 <span class="ot">&lt;-</span> <span class="fu">rowRanges</span>(h3k4me1)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>rd_h3k4me1 <span class="ot">&lt;-</span> rd_h3k4me1[<span class="fu">abs</span>(rd_h3k4me1<span class="sc">$</span>logFC) <span class="sc">&gt;=</span> <span class="fl">0.5</span> <span class="sc">&amp;</span> rd_h3k4me1<span class="sc">$</span>qvalue <span class="sc">&lt;=</span> <span class="fl">0.1</span>]</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(<span class="fu">elementMetadata</span>(rd_h3k4me1)) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"H3K4me1"</span>, <span class="fu">colnames</span>(<span class="fu">elementMetadata</span>(rd_h3k4me1)), <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>overlap <span class="ot">&lt;-</span> <span class="fu">metaGR</span>(<span class="at">gr1 =</span> overlap, <span class="at">gr2 =</span> rd_h3k4me1, <span class="at">minOverlap =</span> <span class="dv">10</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="h3k27me3" class="level3">
<h3 class="anchored" data-anchor-id="h3k27me3">H3K27me3</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>rd_h3k27me3 <span class="ot">&lt;-</span> <span class="fu">rowRanges</span>(h3k27me3)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>rd_h3k27me3 <span class="ot">&lt;-</span> rd_h3k27me3[<span class="fu">abs</span>(rd_h3k27me3<span class="sc">$</span>logFC) <span class="sc">&gt;=</span> <span class="fl">0.5</span> <span class="sc">&amp;</span> rd_h3k27me3<span class="sc">$</span>qvalue <span class="sc">&lt;=</span> <span class="fl">0.1</span>]</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(<span class="fu">elementMetadata</span>(rd_h3k27me3)) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"H3K27me3"</span>, <span class="fu">colnames</span>(<span class="fu">elementMetadata</span>(rd_h3k27me3)), <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>overlap <span class="ot">&lt;-</span> <span class="fu">metaGR</span>(<span class="at">gr1 =</span> overlap, <span class="at">gr2 =</span> rd_h3k27me3, <span class="at">minOverlap =</span> <span class="dv">10</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="h3k27ac" class="level3">
<h3 class="anchored" data-anchor-id="h3k27ac">H3K27ac</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>rd_h3k27ac <span class="ot">&lt;-</span> <span class="fu">rowRanges</span>(h3k27ac)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>rd_h3k27ac <span class="ot">&lt;-</span> rd_h3k27ac[<span class="fu">abs</span>(rd_h3k27ac<span class="sc">$</span>logFC) <span class="sc">&gt;=</span> <span class="fl">0.5</span> <span class="sc">&amp;</span> rd_h3k27ac<span class="sc">$</span>qvalue <span class="sc">&lt;=</span> <span class="fl">0.1</span>]</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(<span class="fu">elementMetadata</span>(rd_h3k27ac)) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"H3K27ac"</span>, <span class="fu">colnames</span>(<span class="fu">elementMetadata</span>(rd_h3k27ac)), <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>overlap <span class="ot">&lt;-</span> <span class="fu">metaGR</span>(<span class="at">gr1 =</span> overlap, <span class="at">gr2 =</span> rd_h3k27ac, <span class="at">minOverlap =</span> <span class="dv">10</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>
<section id="save-data" class="level2">
<h2 class="anchored" data-anchor-id="save-data">Save data</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Add a name field and save the final integrated object</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>overlap<span class="sc">$</span>name <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"peak"</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(overlap), <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(overlap) <span class="ot">&lt;-</span> overlap<span class="sc">$</span>name</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"output"</span>, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(<span class="at">object =</span> overlap, <span class="at">file =</span> <span class="st">"output/overlap_data.rds"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
<section id="snapshot-of-overlamatrix" class="level3">
<h3 class="anchored" data-anchor-id="snapshot-of-overlamatrix">Snapshot of overlaMatrix</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>overlap[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 3 ranges and 109 metadata columns:
         seqnames          ranges strand |        ATAC_annotation ATAC_geneChr
            &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; |            &lt;character&gt;    &lt;integer&gt;
  peak_1     chr1 3670547-3672665      * |       Promoter (&lt;=1kb)            1
  peak_2     chr1 4332510-4332710      * | Intron (ENSMUST00000..            1
  peak_3     chr1 4491755-4492573      * |       Promoter (1-2kb)            1
         ATAC_geneStart ATAC_geneEnd ATAC_geneLength ATAC_geneStrand
              &lt;integer&gt;    &lt;integer&gt;       &lt;integer&gt;       &lt;integer&gt;
  peak_1        3214482      3671498          457017               2
  peak_2        4344146      4360314           16169               2
  peak_3        4492465      4493735            1271               2
         ATAC_geneId    ATAC_transcriptId ATAC_distanceToTSS       ATAC_ENSEMBL
         &lt;character&gt;          &lt;character&gt;          &lt;numeric&gt;        &lt;character&gt;
  peak_1      497097 ENSMUST00000070533.4                  0 ENSMUSG00000051951
  peak_2       19888 ENSMUST00000027032.5              27604 ENSMUSG00000025900
  peak_3       20671 ENSMUST00000191939.1               1162 ENSMUSG00000025902
         ATAC_SYMBOL          ATAC_GENENAME   ATAC_anno ATAC_logFC ATAC_logCPM
         &lt;character&gt;            &lt;character&gt; &lt;character&gt;  &lt;numeric&gt;   &lt;numeric&gt;
  peak_1        Xkr4 X-linked Kx blood gr..    Promoter   0.602984    10.45378
  peak_2         Rp1 retinitis pigmentosa..      Intron   0.855748     5.18424
  peak_3       Sox17 SRY (sex determining..    Promoter  -0.620926     7.13469
            ATAC_F ATAC_PValue ATAC_qvalue      RNA_Row.names RNA_gene_id
         &lt;numeric&gt;   &lt;numeric&gt;   &lt;numeric&gt;        &lt;character&gt; &lt;character&gt;
  peak_1  41.78446 6.58453e-08 4.81983e-07 ENSMUSG00000051951      497097
  peak_2   3.68371 6.13517e-02 9.28785e-02               &lt;NA&gt;        &lt;NA&gt;
  peak_3   9.14522 4.12368e-03 8.38540e-03               &lt;NA&gt;        &lt;NA&gt;
          RNA_SYMBOL           RNA_GENENAME        RNA_ENSEMBL RNA_geneLength
         &lt;character&gt;            &lt;character&gt;        &lt;character&gt;      &lt;integer&gt;
  peak_1        Xkr4 X-linked Kx blood gr.. ENSMUSG00000051951         457017
  peak_2        &lt;NA&gt;                   &lt;NA&gt;               &lt;NA&gt;           &lt;NA&gt;
  peak_3        &lt;NA&gt;                   &lt;NA&gt;               &lt;NA&gt;           &lt;NA&gt;
         RNA_geneChr RNA_geneStart RNA_geneEnd RNA_geneStrand  RNA_geneId
         &lt;character&gt;     &lt;integer&gt;   &lt;integer&gt;    &lt;character&gt; &lt;character&gt;
  peak_1        chr1       3214482     3671498              -      497097
  peak_2        &lt;NA&gt;          &lt;NA&gt;        &lt;NA&gt;           &lt;NA&gt;        &lt;NA&gt;
  peak_3        &lt;NA&gt;          &lt;NA&gt;        &lt;NA&gt;           &lt;NA&gt;        &lt;NA&gt;
             RNA_transcriptId    RNA_anno RNA_logFC RNA_logCPM     RNA_F
                  &lt;character&gt; &lt;character&gt; &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;
  peak_1 ENSMUST00000070533.4        Gene   1.37498    5.62452   33.7077
  peak_2                 &lt;NA&gt;        &lt;NA&gt;        NA         NA        NA
  peak_3                 &lt;NA&gt;        &lt;NA&gt;        NA         NA        NA
         RNA_PValue RNA_qvalue H3K4me3_annotation H3K4me3_geneChr
          &lt;numeric&gt;  &lt;numeric&gt;        &lt;character&gt;       &lt;integer&gt;
  peak_1 0.00224273 0.00628871   Promoter (&lt;=1kb)               1
  peak_2         NA         NA               &lt;NA&gt;            &lt;NA&gt;
  peak_3         NA         NA               &lt;NA&gt;            &lt;NA&gt;
         H3K4me3_geneStart H3K4me3_geneEnd H3K4me3_geneLength
                 &lt;integer&gt;       &lt;integer&gt;          &lt;integer&gt;
  peak_1           3214482         3671498             457017
  peak_2              &lt;NA&gt;            &lt;NA&gt;               &lt;NA&gt;
  peak_3              &lt;NA&gt;            &lt;NA&gt;               &lt;NA&gt;
         H3K4me3_geneStrand H3K4me3_geneId H3K4me3_transcriptId
                  &lt;integer&gt;    &lt;character&gt;          &lt;character&gt;
  peak_1                  2         497097 ENSMUST00000070533.4
  peak_2               &lt;NA&gt;           &lt;NA&gt;                 &lt;NA&gt;
  peak_3               &lt;NA&gt;           &lt;NA&gt;                 &lt;NA&gt;
         H3K4me3_distanceToTSS    H3K4me3_ENSEMBL H3K4me3_SYMBOL
                     &lt;numeric&gt;        &lt;character&gt;    &lt;character&gt;
  peak_1                     0 ENSMUSG00000051951           Xkr4
  peak_2                    NA               &lt;NA&gt;           &lt;NA&gt;
  peak_3                    NA               &lt;NA&gt;           &lt;NA&gt;
               H3K4me3_GENENAME H3K4me3_anno H3K4me3_logFC H3K4me3_logCPM
                    &lt;character&gt;  &lt;character&gt;     &lt;numeric&gt;      &lt;numeric&gt;
  peak_1 X-linked Kx blood gr..     Promoter      0.607114        9.26749
  peak_2                   &lt;NA&gt;         &lt;NA&gt;            NA             NA
  peak_3                   &lt;NA&gt;         &lt;NA&gt;            NA             NA
         H3K4me3_F H3K4me3_PValue H3K4me3_qvalue H3K4me1_annotation
         &lt;numeric&gt;      &lt;numeric&gt;      &lt;numeric&gt;        &lt;character&gt;
  peak_1   135.443    4.65079e-31    1.28968e-29               &lt;NA&gt;
  peak_2        NA             NA             NA               &lt;NA&gt;
  peak_3        NA             NA             NA               &lt;NA&gt;
         H3K4me1_geneChr H3K4me1_geneStart H3K4me1_geneEnd H3K4me1_geneLength
               &lt;integer&gt;         &lt;integer&gt;       &lt;integer&gt;          &lt;integer&gt;
  peak_1            &lt;NA&gt;              &lt;NA&gt;            &lt;NA&gt;               &lt;NA&gt;
  peak_2            &lt;NA&gt;              &lt;NA&gt;            &lt;NA&gt;               &lt;NA&gt;
  peak_3            &lt;NA&gt;              &lt;NA&gt;            &lt;NA&gt;               &lt;NA&gt;
         H3K4me1_geneStrand H3K4me1_geneId H3K4me1_transcriptId
                  &lt;integer&gt;    &lt;character&gt;          &lt;character&gt;
  peak_1               &lt;NA&gt;           &lt;NA&gt;                 &lt;NA&gt;
  peak_2               &lt;NA&gt;           &lt;NA&gt;                 &lt;NA&gt;
  peak_3               &lt;NA&gt;           &lt;NA&gt;                 &lt;NA&gt;
         H3K4me1_distanceToTSS H3K4me1_ENSEMBL H3K4me1_SYMBOL H3K4me1_GENENAME
                     &lt;numeric&gt;     &lt;character&gt;    &lt;character&gt;      &lt;character&gt;
  peak_1                    NA            &lt;NA&gt;           &lt;NA&gt;             &lt;NA&gt;
  peak_2                    NA            &lt;NA&gt;           &lt;NA&gt;             &lt;NA&gt;
  peak_3                    NA            &lt;NA&gt;           &lt;NA&gt;             &lt;NA&gt;
         H3K4me1_anno H3K4me1_logFC H3K4me1_logCPM H3K4me1_F H3K4me1_PValue
          &lt;character&gt;     &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt;      &lt;numeric&gt;
  peak_1         &lt;NA&gt;            NA             NA        NA             NA
  peak_2         &lt;NA&gt;            NA             NA        NA             NA
  peak_3         &lt;NA&gt;            NA             NA        NA             NA
         H3K4me1_qvalue H3K27me3_annotation H3K27me3_geneChr H3K27me3_geneStart
              &lt;numeric&gt;         &lt;character&gt;        &lt;integer&gt;          &lt;integer&gt;
  peak_1             NA                &lt;NA&gt;             &lt;NA&gt;               &lt;NA&gt;
  peak_2             NA                &lt;NA&gt;             &lt;NA&gt;               &lt;NA&gt;
  peak_3             NA    Promoter (&lt;=1kb)                1            4492465
         H3K27me3_geneEnd H3K27me3_geneLength H3K27me3_geneStrand
                &lt;integer&gt;           &lt;integer&gt;           &lt;integer&gt;
  peak_1             &lt;NA&gt;                &lt;NA&gt;                &lt;NA&gt;
  peak_2             &lt;NA&gt;                &lt;NA&gt;                &lt;NA&gt;
  peak_3          4493735                1271                   2
         H3K27me3_geneId H3K27me3_transcriptId H3K27me3_distanceToTSS
             &lt;character&gt;           &lt;character&gt;              &lt;numeric&gt;
  peak_1            &lt;NA&gt;                  &lt;NA&gt;                     NA
  peak_2            &lt;NA&gt;                  &lt;NA&gt;                     NA
  peak_3           20671  ENSMUST00000191939.1                      0
           H3K27me3_ENSEMBL H3K27me3_SYMBOL      H3K27me3_GENENAME
                &lt;character&gt;     &lt;character&gt;            &lt;character&gt;
  peak_1               &lt;NA&gt;            &lt;NA&gt;                   &lt;NA&gt;
  peak_2               &lt;NA&gt;            &lt;NA&gt;                   &lt;NA&gt;
  peak_3 ENSMUSG00000025902           Sox17 SRY (sex determining..
         H3K27me3_anno H3K27me3_logFC H3K27me3_logCPM H3K27me3_F
           &lt;character&gt;      &lt;numeric&gt;       &lt;numeric&gt;  &lt;numeric&gt;
  peak_1          &lt;NA&gt;             NA              NA         NA
  peak_2          &lt;NA&gt;             NA              NA         NA
  peak_3      Promoter        -1.0597         9.04668    87.0802
         H3K27me3_PValue H3K27me3_qvalue H3K27ac_annotation H3K27ac_geneChr
               &lt;numeric&gt;       &lt;numeric&gt;        &lt;character&gt;       &lt;integer&gt;
  peak_1              NA              NA   Promoter (&lt;=1kb)               1
  peak_2              NA              NA               &lt;NA&gt;            &lt;NA&gt;
  peak_3     1.32026e-20     5.28858e-18               &lt;NA&gt;            &lt;NA&gt;
         H3K27ac_geneStart H3K27ac_geneEnd H3K27ac_geneLength
                 &lt;integer&gt;       &lt;integer&gt;          &lt;integer&gt;
  peak_1           3214482         3671498             457017
  peak_2              &lt;NA&gt;            &lt;NA&gt;               &lt;NA&gt;
  peak_3              &lt;NA&gt;            &lt;NA&gt;               &lt;NA&gt;
         H3K27ac_geneStrand H3K27ac_geneId H3K27ac_transcriptId
                  &lt;integer&gt;    &lt;character&gt;          &lt;character&gt;
  peak_1                  2         497097 ENSMUST00000070533.4
  peak_2               &lt;NA&gt;           &lt;NA&gt;                 &lt;NA&gt;
  peak_3               &lt;NA&gt;           &lt;NA&gt;                 &lt;NA&gt;
         H3K27ac_distanceToTSS    H3K27ac_ENSEMBL H3K27ac_SYMBOL
                     &lt;numeric&gt;        &lt;character&gt;    &lt;character&gt;
  peak_1                     0 ENSMUSG00000051951           Xkr4
  peak_2                    NA               &lt;NA&gt;           &lt;NA&gt;
  peak_3                    NA               &lt;NA&gt;           &lt;NA&gt;
               H3K27ac_GENENAME H3K27ac_anno H3K27ac_logFC H3K27ac_logCPM
                    &lt;character&gt;  &lt;character&gt;     &lt;numeric&gt;      &lt;numeric&gt;
  peak_1 X-linked Kx blood gr..     Promoter       1.25269        8.81735
  peak_2                   &lt;NA&gt;         &lt;NA&gt;            NA             NA
  peak_3                   &lt;NA&gt;         &lt;NA&gt;            NA             NA
         H3K27ac_F H3K27ac_PValue H3K27ac_qvalue        name
         &lt;numeric&gt;      &lt;numeric&gt;      &lt;numeric&gt; &lt;character&gt;
  peak_1   151.496    1.67395e-34     4.0813e-32      peak_1
  peak_2        NA             NA             NA      peak_2
  peak_3        NA             NA             NA      peak_3
  -------
  seqinfo: 21 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="normalized-metrics-for-plotting" class="level1">
<h1>Normalized metrics for plotting</h1>
<p>Recommendations based on ENCODE Project Consortium and other literature:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 26%">
<col style="width: 26%">
<col style="width: 46%">
</colgroup>
<thead>
<tr class="header">
<th>Assays</th>
<th>Typical Window Size (± kb)</th>
<th>Reason / Notes</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>ATAC-seq</strong></td>
<td>±1 kb</td>
<td>Narrow peaks, accessibility centered on summit</td>
</tr>
<tr class="even">
<td><strong>H3K4me3</strong></td>
<td>±1 kb to ±2 kb</td>
<td>Promoter mark, sharp and narrow peaks</td>
</tr>
<tr class="odd">
<td><strong>H3K4me1</strong></td>
<td>±2 kb to ±5 kb</td>
<td>Enhancer-associated, moderately broad</td>
</tr>
<tr class="even">
<td><strong>H3K27me3</strong></td>
<td>±5 kb to ±10 kb</td>
<td>Broad repressive domains</td>
</tr>
<tr class="odd">
<td><strong>H3K27ac</strong></td>
<td>±2 kb to ±5 kb</td>
<td>Active enhancers/promoters, broader peaks</td>
</tr>
</tbody>
</table>
<p>We can check the <code>mean</code> and <code>median</code> of each histone marks and make the <code>normalizedMatrix</code> accordingly.</p>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate mean and median widths for each dataset</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>summary_table <span class="ot">&lt;-</span> <span class="fu">tibble</span>(</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Assay =</span> <span class="fu">c</span>(<span class="st">"ATAC-seq"</span>, <span class="st">"H3K4me3"</span>, <span class="st">"H3K4me1"</span>, <span class="st">"H3K27me3"</span>, <span class="st">"H3K27ac"</span>),</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Mean_Width =</span> <span class="fu">c</span>(</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mean</span>(<span class="fu">width</span>(rd_atac)),</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mean</span>(<span class="fu">width</span>(rd_h3k4me3)),</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mean</span>(<span class="fu">width</span>(rd_h3k4me1)),</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mean</span>(<span class="fu">width</span>(rd_h3k27me3)),</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">mean</span>(<span class="fu">width</span>(rd_h3k27ac))</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">Median_Width =</span> <span class="fu">c</span>(</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">median</span>(<span class="fu">width</span>(rd_atac)),</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">median</span>(<span class="fu">width</span>(rd_h3k4me3)),</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">median</span>(<span class="fu">width</span>(rd_h3k4me1)),</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">median</span>(<span class="fu">width</span>(rd_h3k27me3)),</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">median</span>(<span class="fu">width</span>(rd_h3k27ac))</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">Min_Width =</span> <span class="fu">c</span>(</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">min</span>(<span class="fu">width</span>(rd_atac)),</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">min</span>(<span class="fu">width</span>(rd_h3k4me3)),</span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">min</span>(<span class="fu">width</span>(rd_h3k4me1)),</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">min</span>(<span class="fu">width</span>(rd_h3k27me3)),</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">min</span>(<span class="fu">width</span>(rd_h3k27ac))</span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>  ),</span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>  <span class="at">Max_Width =</span> <span class="fu">c</span>(</span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">max</span>(<span class="fu">width</span>(rd_atac)),</span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">max</span>(<span class="fu">width</span>(rd_h3k4me3)),</span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">max</span>(<span class="fu">width</span>(rd_h3k4me1)),</span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>    <span class="fu">max</span>(<span class="fu">width</span>(rd_h3k27me3)),</span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a>    <span class="fu">max</span>(<span class="fu">width</span>(rd_h3k27ac))</span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(summary_table)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 5
  Assay    Mean_Width Median_Width Min_Width Max_Width
  &lt;chr&gt;         &lt;dbl&gt;        &lt;dbl&gt;     &lt;int&gt;     &lt;int&gt;
1 ATAC-seq       956.          932       151      3490
2 H3K4me3       1316.          749       191     18723
3 H3K4me1        791.          563       171     12590
4 H3K27me3      1170.          568       176     23571
5 H3K27ac        914.          618       171     22276</code></pre>
</div>
</div>
<p>As we restricted the data for this course to just 2 chromosomes, maximum peaks are around 1kb wide. Hence, we will go ahead with <code>+/- 1kb</code> windows for our analysis.</p>
<section id="taking-1000-bp-around-the-mid-of-atac-peaks" class="level2">
<h2 class="anchored" data-anchor-id="taking-1000-bp-around-the-mid-of-atac-peaks">Taking 1000 bp around the mid of ATAC-peaks</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract midpoints of peaks to create windows for visualization</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>mid_peaks <span class="ot">&lt;-</span> <span class="fu">gr.mid</span>(overlap)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="atac" class="level2">
<h2 class="anchored" data-anchor-id="atac">ATAC</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load ATAC bigWig files</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>atac_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data"</span>, <span class="at">pattern =</span> <span class="st">"ATAC"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(atac_files) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.bw"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> <span class="fu">basename</span>(atac_files))</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>atac_bw <span class="ot">&lt;-</span> <span class="fu">lapply</span>(atac_files, <span class="cf">function</span>(x){</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  a <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(x)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>  a <span class="ot">&lt;-</span> a[<span class="fu">seqnames</span>(a) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>)]</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  a</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize signal around mid_peaks</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>mat_AS <span class="ot">&lt;-</span> <span class="fu">lapply</span>(atac_bw, <span class="at">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">normalizeToMatrix</span>(x, mid_peaks,</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">extend =</span> <span class="dv">1000</span>,</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">value_column =</span> <span class="st">"score"</span>,</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">include_target =</span> <span class="cn">TRUE</span>,</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_mode =</span> <span class="st">"w0"</span>,</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">w =</span> <span class="dv">20</span>, </span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>    <span class="at">smooth =</span> T,</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>    <span class="at">background =</span> <span class="dv">0</span></span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(mat_AS, <span class="at">file =</span> <span class="st">"output/mat_atac.rds"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="chip" class="level2">
<h2 class="anchored" data-anchor-id="chip">ChIP</h2>
<section id="h3k4me3-1" class="level3">
<h3 class="anchored" data-anchor-id="h3k4me3-1">H3K4me3</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>h3k4me3_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data"</span>, <span class="at">pattern =</span> <span class="st">"H3K4me3"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(h3k4me3_files) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.bw"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> <span class="fu">basename</span>(h3k4me3_files))</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>h3k4me3_bw <span class="ot">&lt;-</span> <span class="fu">lapply</span>(h3k4me3_files, <span class="cf">function</span>(x){</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>  a <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(x)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  a <span class="ot">&lt;-</span> a[<span class="fu">seqnames</span>(a) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>)]</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  a</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>mat_h3k4me3 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(h3k4me3_bw, <span class="at">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">normalizeToMatrix</span>(x, mid_peaks,</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">extend =</span> <span class="dv">1000</span>,</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">value_column =</span> <span class="st">"score"</span>,</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">include_target =</span> <span class="cn">TRUE</span>,</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_mode =</span> <span class="st">"w0"</span>,</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">w =</span> <span class="dv">20</span>,</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">smooth =</span> T,</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">background =</span> <span class="dv">0</span></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(mat_h3k4me3, <span class="at">file =</span> <span class="st">"output/mat_h3k4me3.rds"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="h3k4me1-1" class="level3">
<h3 class="anchored" data-anchor-id="h3k4me1-1">H3K4me1</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>h3k4me1_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data"</span>, <span class="at">pattern =</span> <span class="st">"H3K4me1"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(h3k4me1_files) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.bw"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> <span class="fu">basename</span>(h3k4me1_files))</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>h3k4me1_bw <span class="ot">&lt;-</span> <span class="fu">lapply</span>(h3k4me1_files, <span class="cf">function</span>(x){</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  a <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(x)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  a <span class="ot">&lt;-</span> a[<span class="fu">seqnames</span>(a) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>)]</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  a</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>mat_h3k4me1 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(h3k4me1_bw, <span class="at">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">normalizeToMatrix</span>(x, mid_peaks,</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">extend =</span> <span class="dv">1000</span>,</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">value_column =</span> <span class="st">"score"</span>,</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">include_target =</span> <span class="cn">TRUE</span>,</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_mode =</span> <span class="st">"w0"</span>,</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">w =</span> <span class="dv">20</span>,</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">smooth =</span> T,</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">background =</span> <span class="dv">0</span></span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(mat_h3k4me1, <span class="at">file =</span> <span class="st">"output/mat_h3k4me1.rds"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="h27k4me3" class="level3">
<h3 class="anchored" data-anchor-id="h27k4me3">H27K4me3</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>h3k27me3_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data"</span>, <span class="at">pattern =</span> <span class="st">"H3K27me3"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(h3k27me3_files) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.bw"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> <span class="fu">basename</span>(h3k27me3_files))</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>h3k27me3_bw <span class="ot">&lt;-</span> <span class="fu">lapply</span>(h3k27me3_files, <span class="cf">function</span>(x){</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  a <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(x)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  a <span class="ot">&lt;-</span> a[<span class="fu">seqnames</span>(a) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>)]</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  a</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>mat_h3k27me3 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(h3k27me3_bw, <span class="at">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">normalizeToMatrix</span>(x, mid_peaks,</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">extend =</span> <span class="dv">1000</span>,</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">value_column =</span> <span class="st">"score"</span>,</span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">include_target =</span> <span class="cn">TRUE</span>,</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_mode =</span> <span class="st">"w0"</span>,</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">w =</span> <span class="dv">20</span>,</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">smooth =</span> T,</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">background =</span> <span class="dv">0</span></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(mat_h3k27me3, <span class="at">file =</span> <span class="st">"output/mat_h3k27me3.rds"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="h3k27ac-1" class="level3">
<h3 class="anchored" data-anchor-id="h3k27ac-1">H3K27ac</h3>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>h3k27ac_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data"</span>, <span class="at">pattern =</span> <span class="st">"H3K27ac"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(h3k27ac_files) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.bw"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> <span class="fu">basename</span>(h3k27ac_files))</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>h3k27ac_bw <span class="ot">&lt;-</span> <span class="fu">lapply</span>(h3k27ac_files, <span class="cf">function</span>(x){</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  a <span class="ot">&lt;-</span> rtracklayer<span class="sc">::</span><span class="fu">import</span>(x)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  a <span class="ot">&lt;-</span> a[<span class="fu">seqnames</span>(a) <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>)]</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>  a</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>mat_h3k27ac <span class="ot">&lt;-</span> <span class="fu">lapply</span>(h3k27ac_bw, <span class="at">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">normalizeToMatrix</span>(x, mid_peaks,</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>    <span class="at">extend =</span> <span class="dv">1000</span>,</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>    <span class="at">value_column =</span> <span class="st">"score"</span>,</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>    <span class="at">include_target =</span> <span class="cn">TRUE</span>,</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>    <span class="at">mean_mode =</span> <span class="st">"w0"</span>,</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>    <span class="at">w =</span> <span class="dv">20</span>,</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>    <span class="at">smooth =</span> T,</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>    <span class="at">background =</span> <span class="dv">0</span></span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(mat_h3k27ac, <span class="at">file =</span> <span class="st">"output/mat_h3k27ac.rds"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
</section>
<section id="rna-1" class="level2">
<h2 class="anchored" data-anchor-id="rna-1">RNA</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Match overlapping peak names with RNA-seq gene names</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>tmp <span class="ot">&lt;-</span> <span class="fu">elementMetadata</span>(overlap)[,<span class="fu">c</span>(<span class="st">"RNA_Row.names"</span>, <span class="st">"name"</span>)]</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Get logCPM matrix</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">assay</span>(rna, <span class="st">"logCPM"</span>)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">assay</span>(rna, <span class="st">"logCPM"</span>) <span class="sc">-</span> <span class="fu">rowMeans</span>(</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">assay</span>(rna, <span class="st">"logCPM"</span>)[, <span class="fu">grep</span>(<span class="at">pattern =</span> <span class="st">"11half"</span>, <span class="at">x =</span> <span class="fu">colnames</span>(<span class="fu">assay</span>(rna, <span class="st">"logCPM"</span>)),</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>                              <span class="at">value =</span> T)]</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge gene expression values with peaks</span></span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>mat_RNA <span class="ot">&lt;-</span> <span class="fu">merge</span>(tmp, counts, <span class="at">by.x =</span> <span class="st">"RNA_Row.names"</span>, <span class="at">by.y =</span> <span class="st">"row.names"</span>, <span class="at">all.x =</span> T)</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Set rownames to peak names and cleanup</span></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(mat_RNA) <span class="ot">&lt;-</span> mat_RNA<span class="sc">$</span>name</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>mat_RNA <span class="ot">&lt;-</span> mat_RNA[,<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)]</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(mat_RNA) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">".tsv.gz"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> <span class="fu">colnames</span>(mat_RNA))</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>mat_RNA <span class="ot">&lt;-</span> <span class="fu">data.matrix</span>(mat_RNA)</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>mat_RNA <span class="ot">&lt;-</span> mat_RNA[<span class="fu">names</span>(overlap),]</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Save data</span></span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(mat_RNA, <span class="at">file =</span> <span class="st">"output/mat_rna.rds"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="wgbs" class="level2">
<h2 class="anchored" data-anchor-id="wgbs">WGBS</h2>
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Bisulfite seq coverage files after methylation call</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>bs_files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(<span class="st">"data"</span>, <span class="at">pattern =</span> <span class="st">"WGBS"</span>, <span class="at">full.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(bs_files) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">.bed.gz"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> <span class="fu">basename</span>(bs_files))</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Reading coverage</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>bs_cov <span class="ot">&lt;-</span> <span class="fu">lapply</span>(bs_files, <span class="cf">function</span>(x) {</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">GRanges</span>(  </span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">fread</span>(</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>      <span class="at">input =</span> x,</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">sep =</span> <span class="st">" "</span>, <span class="at">quote =</span> F, <span class="at">stringsAsFactors =</span> F,</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">data.table =</span> <span class="cn">FALSE</span>, <span class="at">nThread =</span> parallel<span class="sc">::</span><span class="fu">detectCores</span>(), <span class="at">showProgress =</span> F,</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>      <span class="at">col.names =</span> <span class="fu">c</span>(<span class="st">"seqnames"</span>, <span class="st">"start"</span>, <span class="st">"end"</span>, <span class="st">"cov"</span>, <span class="st">"Me"</span>, <span class="st">"Un"</span>, <span class="st">"meth"</span>)</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalized matrics</span></span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>mat_bs <span class="ot">&lt;-</span> <span class="fu">lapply</span>(bs_cov, <span class="at">FUN =</span> <span class="cf">function</span>(x) {</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">normalizeToMatrix</span>(x, mid_peaks,</span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">extend =</span> <span class="dv">1000</span>,</span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a>      <span class="at">value_column =</span> <span class="st">"meth"</span>,</span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">include_target =</span> <span class="cn">TRUE</span>,</span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a>      <span class="at">smooth =</span> <span class="cn">TRUE</span>,</span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">mean_mode =</span> <span class="st">"absolute"</span>,</span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a>      <span class="at">background =</span> <span class="cn">NA</span></span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb22-27"><a href="#cb22-27" aria-hidden="true" tabindex="-1"></a>})</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>All signal values are within [0, 1], so we assume it is methylation
signal. Automatically set limit [0, 1] to the smoothed values. If this
is not the case, set argument `limit = NA` in the function to remove
the limits. Set `verbose = FALSE` to turn off this message.</code></pre>
</div>
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(mat_bs, <span class="at">file =</span> <span class="st">"output/mat_bs.rds"</span>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
</div>
</section>
<section id="question-1" class="level2">
<h2 class="anchored" data-anchor-id="question-1">Question 1</h2>
<p><strong>Make an <code>EnrichedHeatmap</code></strong> of methylation data.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center collapsed" data-bs-toggle="collapse" data-bs-target=".callout-1-contents" aria-controls="callout-1" aria-expanded="false" aria-label="Toggle callout">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
<div class="callout-btn-toggle d-inline-block border-0 py-1 ps-1 pe-0 float-end"><i class="callout-toggle"></i></div>
</div>
<div id="callout-1" class="callout-1-contents callout-collapse collapse">
<div class="callout-body-container callout-body">
<div class="cell">
<div class="code-copy-outer-scaffold"><div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">EnrichedHeatmap</span>(</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">mat =</span> mat_bs<span class="sc">$</span>WGBS_11half,    <span class="co"># normalized matrix</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">name =</span> <span class="st">"E11.5"</span>,              <span class="co"># Name for the plot</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">width =</span> <span class="fu">unit</span>(<span class="dv">4</span>, <span class="st">"cm"</span>),       <span class="co"># Width of the heatmap</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">height =</span> <span class="fu">unit</span>(<span class="dv">8</span>, <span class="st">"cm"</span>)      <span class="co"># Height of the heatmap</span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>) <span class="sc">+</span> <span class="fu">EnrichedHeatmap</span>(</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">mat =</span> mat_bs<span class="sc">$</span>WGBS_15half,    <span class="co"># normalized matrix</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">name =</span> <span class="st">"E15.5"</span>,              <span class="co"># Name for the plot</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">width =</span> <span class="fu">unit</span>(<span class="dv">4</span>, <span class="st">"cm"</span>),       <span class="co"># Width of the heatmap</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">height =</span> <span class="fu">unit</span>(<span class="dv">8</span>, <span class="st">"cm"</span>)      <span class="co"># Height of the heatmap</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></div>
<div class="cell-output cell-output-stderr">
<pre><code>The automatically generated colors map from the 1^st and 99^th of the
values in the matrix. There are outliers in the matrix whose patterns
might be hidden by this color mapping. You can manually set the color
to `col` argument.

Use `suppressMessages()` to turn off this message.
The automatically generated colors map from the 1^st and 99^th of the
values in the matrix. There are outliers in the matrix whose patterns
might be hidden by this color mapping. You can manually set the color
to `col` argument.

Use `suppressMessages()` to turn off this message.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="day2-1_overlap_matrix_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Important
</div>
</div>
<div class="callout-body-container callout-body">
<p>We have not performed differential analysis for DNA methylation data. If you want, you can. Here, we are using DNA methylation as an observartory mark.</p>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
      const outerScaffold = trigger.parentElement.cloneNode(true);
      const codeEl = outerScaffold.querySelector('code');
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp("https:\/\/sib-swiss\.github\.io\/multiomics-biological-integration-training\/");
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>